File: srs.js
/**
 * SM-2 spaced repetition algorithm.
 *
 * @typedef {{ interval: number, repetitions: number, easeFactor: number, nextReview: number }} SRSItem
 * @typedef {Object.<string, SRSItem>} SRSState
 */

const DEFAULT_EASE = 2.5;

/**
 * Compute next SRS interval for an item.
 * @param {SRSItem} item
 * @param {number} quality - 0 (blackout) to 5 (perfect)
 * @returns {SRSItem} updated item (mutated in place and returned)
 */
export function sm2(item, quality) {
    if (quality < 3) {
        item.repetitions = 0;
        item.interval = 1;
    } else {
        if (item.repetitions === 0) {
            item.interval = 1;
        } else if (item.repetitions === 1) {
            item.interval = 3;
        } else {
            item.interval = Math.round(item.interval * item.easeFactor);
        }
        item.repetitions++;
        item.easeFactor = Math.max(
            1.3,
            item.easeFactor + 0.1 - (5 - quality) * (0.08 + (5 - quality) * 0.02)
        );
    }
    item.nextReview = Date.now() + item.interval * 86_400_000;
    return item;
}

/**
 * @param {string} storageKey
 * @returns {SRSState}
 */
export function loadSRSState(storageKey) {
    try {
        return JSON.parse(localStorage.getItem(storageKey) || '{}');
    } catch {
        return {};
    }
}

/**
 * @param {string} storageKey
 * @param {SRSState} state
 */
export function saveSRSState(storageKey, state) {
    localStorage.setItem(storageKey, JSON.stringify(state));
}

/**
 * Get or initialise the SRS record for a single item key.
 * @param {SRSState} state
 * @param {string} key
 * @returns {SRSItem}
 */
export function getItemSRS(state, key) {
    if (!state[key]) {
        state[key] = { interval: 0, repetitions: 0, easeFactor: DEFAULT_EASE, nextReview: 0 };
    }
    return state[key];
}

/**
 * Apply a SM-2 quality score to an item and persist the state.
 * @param {SRSState} state
 * @param {string} key
 * @param {number} quality - 0–5
 * @param {string} storageKey - localStorage key for persistence
 * @returns {SRSItem}
 */
export function updateItemSRS(state, key, quality, storageKey) {
    const item = getItemSRS(state, key);
    sm2(item, quality);
    saveSRSState(storageKey, state);
    return item;
}

/**
 * Return all keys whose nextReview timestamp is in the past (due now).
 * @param {SRSState} state
 * @returns {string[]}
 */
export function getDueItems(state) {
    const now = Date.now();
    return Object.keys(state).filter(k => state[k].nextReview > 0 && state[k].nextReview <= now);
}

/**
 * @param {SRSState} state
 * @param {string} key
 * @returns {boolean}
 */
export function isItemDue(state, key) {
    const item = state[key];
    return item ? (item.nextReview > 0 && item.nextReview <= Date.now()) : false;
}

File: android/app/src/main/java/com/ktouchie/quizportugues/srs/Srs.kt
package com.ktouchie.quizportugues.srs

/**
 * SM-2 spaced repetition algorithm — Kotlin port of `srs.js`. This is a behavioral port, not
 * shared code (see docs/MOBILE_APP_SPEC.md §7): Kotlin can't consume the web app's JS, so this is
 * a deliberate line-for-line translation of the same algorithm, kept in sync by the tests in
 * SrsTest.kt mirroring `tests/srs.test.js`'s cases.
 *
 * Where `srs.js` mutates its item object in place, this port returns a new [SrsRecord] instead —
 * an idiomatic Kotlin adjustment that keeps the function pure, not a behavioral change.
 */

const val DEFAULT_EASE = 2.5
private const val DAY_MILLIS = 86_400_000L

/**
 * One item's spaced-repetition state. Mirrors `srs.js`'s `SRSItem` shape 1:1 — this is also the
 * shape persisted in the Room `srs_records` table (see the Room schema task).
 */
data class SrsRecord(
    val interval: Int = 0,
    val repetitions: Int = 0,
    val easeFactor: Double = DEFAULT_EASE,
    val nextReview: Long = 0L,
)

/**
 * Compute the next SRS state for an item after answering it.
 *
 * @param quality 0 (blackout) to 5 (perfect) — the app only ever produces 4 (first-try correct),
 *   2 (correct after a mistake), or 0 (wrong), matching `quiz_base.js`'s scoring today.
 * @param now injectable for deterministic tests; defaults to the real clock.
 */
fun sm2(item: SrsRecord, quality: Int, now: Long = System.currentTimeMillis()): SrsRecord {
    // Ease factor is always updated, even on failed recalls.
    val newEase = (item.easeFactor + 0.1 - (5 - quality) * (0.08 + (5 - quality) * 0.02))
        .coerceAtLeast(1.3)

    val (newRepetitions, newInterval) = if (quality < 3) {
        // Failed or barely recalled — reset schedule, item comes back tomorrow.
        0 to 1
    } else {
        val interval = when (item.repetitions) {
            0 -> 1
            1 -> 3
            else -> Math.round(item.interval * item.easeFactor).toInt()
        }
        (item.repetitions + 1) to interval
    }

    return item.copy(
        easeFactor = newEase,
        repetitions = newRepetitions,
        interval = newInterval,
        nextReview = now + newInterval * DAY_MILLIS,
    )
}

/** Mirrors `isItemDue()` in srs.js. */
fun isDue(record: SrsRecord, now: Long = System.currentTimeMillis()): Boolean =
    record.nextReview > 0 && record.nextReview <= now

/**
 * Mirrors `getDueItems()` in srs.js — filters a map of item-id -> record down to the ids that are
 * due now. The Room repository layer (see the Room schema/DAO tasks) is expected to query for
 * this directly rather than loading everything into memory; this is kept as a plain utility over
 * a [Map] for unit testing and for any in-memory session use.
 */
fun dueItemIds(records: Map<String, SrsRecord>, now: Long = System.currentTimeMillis()): List<String> =
    records.filterValues { isDue(it, now) }.keys.toList()
